django/utils/functional.py

```diff
@@ -238,11 +237,12 @@ def decorator(func):
 
         @wraps(func)
         def wrapper(*args, **kwargs):
-            if any(
-                isinstance(arg, Promise)
-                for arg in itertools.chain(args, kwargs.values())
-            ):
-                return lazy_func(*args, **kwargs)
+            for arg in args:
+                if isinstance(arg, Promise):
+                    return lazy_func(*args, **kwargs)
+            for arg in kwargs.values():
+                if isinstance(arg, Promise):
+                    return lazy_func(*args, **kwargs)
```

Comment:
It'd be interesting to see some numbers for doing it this way[^1], as it may be 'good enough' to obviate #14849 entirely, which would be nice from a simplicity point of view.

[^1]: and indeed I'll try and check the numbers against my own at some point.

Reply:
I think it's still slow, so something like https://github.com/django/django/pull/14849 would still help. Like if we know the function doesn't accept `kwargs`, we could completely avoid calling `kwargs.values()` (`django.utils.html.escape()` for example.

Or maybe `escape()` gets called so much that it might be worth just hand-coding the `keep_lazy` code. There's only one arg to check.

Here's https://github.com/gaogaotiantian/viztracer for `conditional_escape` calling `escape` with this patch applied:

<img width="1163" alt="image" src="https://user-images.githubusercontent.com/129234/155222258-0ef4ae5a-4e4d-43f2-a48f-bba441e41f41.png">

Yeah, maybe the function itself could check its args for `Promise` and call `lazy(itself)` if `Promise` found. That way functions could just be hand-optimized and they wouldn't need to be wrapped at all if no promises are passed in, which means fewer function calls.

Example for `django.utils.html.escape()`, instead of `@keep_lazy(SafeString)` wrapper, just do this:
```
def escape(text):
    if isinstance(text, Promise):
        return _lazy_escape(text)

    return SafeString(html.escape(str(text)))

_lazy_escape = lazy(escape, SafeString)
```

It means if a lazy value is passed in, there's an extra `Promise` check, but at least it's faster for the non-promise case.

Here's viztracer for this example, you can see there's one fewer function call, and the dark-yellow `kwargs.values()` call is gone.

<img width="1143" alt="image" src="https://user-images.githubusercontent.com/129234/155226074-5d976882-465a-4361-9776-09c3a7fb2a31.png">